### ISSUE TITLE
Thumbnail Rotation

### ISSUE THREAD

**Issue by miqrogroove** [fisharebest/webtrees#3553]
In v2.0.9 I just noticed a photo thumbnail that's not rotated properly.  When I click on it, the full size image is correctly rotated.

It looks like I uploaded that photo in June, so I was probably running 2.0.3 or 2.0.5 at the time.

For a test, I just uploaded the file again and saw the same problem.

**Comment by fisharebest** [comment]
Does your server use `GD` or `Imagick` to create thumbnails?

**Comment by miqrogroove** [comment]
PHP 7.4.3

EXIF Support | enabled
-- | --
Supported EXIF Version | 0220


GD Support | enabled
-- | --
GD headers Version | 2.2.5
GD library Version | 2.2.5

imagick module version | 3.4.4
-- | --

**Comment by HonkXL** [comment]
I had the same problem: some of my older thumbnails where in a wrong rotation, while image rotation is ok. EXIF module also installed.
I solved it by downloading the image by FTP, rotating it in gimp and uploading it again. No really good way, but I only had a few of this images.

**Comment by brianegge** [comment]
Same issue. Both GD and imagick are installed.
```

EXIF Support | enabled
-- | --
Supported EXIF Version | 0220
Supported filetypes | JPEG, TIFF
Multibyte decoding support using mbstring | enabled
Extended EXIF tag formats | Canon, Casio, Fujifilm, Nikon, Olympus, Samsung, Panasonic, DJI, Sony, Pentax, Minolta, Sigma, Foveon, Kyocera, Ricoh, AGFA, Epson

```

**Comment by fisharebest** [comment]
webtrees uses the "industry standard" image handling library `intervention`.

webtrees uses its `orientate()` method to apply auto-rotation of images using EXIF data.
You can see the documentation here: <http://image.intervention.io/api/orientate>

If this library cannot auto-rotate the image, then there is probably nothing I can do.

You could, perhaps, add `throw $ex;` here,

https://github.com/fisharebest/webtrees/blob/2.0.15/app/Factories/ImageFactory.php#L381

which might provide some information as to why this image cannot be rotated.

**Comment by brianegge** [comment]
According to the intervention docs, the it must have a file path in order to read the exif data. In ImageFactories all the images are read through a stream.

**Comment by brianegge** [comment]
https://github.com/Intervention/image/issues/745

**Comment by fisharebest** [comment]
> In ImageFactories all the images are read through a stream.

All file handling is via an abstraction layer - <https://flysystem.thephpleague.com>.

This is to allow us to use a remote filesystem for storage of media files (e.g. DropBox, OneDrive, AWS S3, etc.).

So media files will always be read via a stream.

**Comment by miqrogroove** [comment]
Then copy the stream to a temporary file and use the full path.  There's no excuse for bad thumbnail rotation.  😄

**Comment by brianegge** [comment]
If the underlying resource is a file, then you can get the path, even through the fly system. However, I think the chroot adapter doesn't forward the path() function. It might be possible to pass the bytes buffer instead of a stream. It sounds like a random access buffer may be able to return the exif data.

**Comment by fisharebest** [comment]
> If the underlying resource is a file, then you can get the path, even through the fly system.

In flysystem 1.0, it was possible to access the adapter.  So, if it was a LocalFilesystem adapter, you could then access the file directly.  In flysystem 2.0 (and hence in webtrees 2.1.0 onwards), this option was removed.

IMHO, this is a bug in the intervention library.

**Comment by brianegge** [comment]
Loading a jpeg from a stream is of course efficient. I think exif can work if the file is read into a byte array.

**Comment by mstimberg** [comment]
This appears to be fixed in intervention 2.6 (Intervention/image#1054), based on the php 7.2 feature to [read exif data from streams](https://www.php.net/manual/en/function.exif-read-data.php#refsect1-function.exif-read-data-changelog).

**Comment by fisharebest** [comment]
@mstimberg - thanks for reporting this.  Both the `main` and `2.0` branches have been updated to use `2.6.1`, so this fix will be included in the next release.

**Comment by HonkXL** [comment]
seems to work now - but is there a easy way to fix the wrong thumbnails?

**Comment by fisharebest** [comment]
`2.0.17` has now been released, so this fix is now available.

**Comment by miqrogroove** [comment]
I can't seem to find the thumbnail cache.  I have a custom data folder path.  It's reading media normally, but all the cache dirs seem empty.  I'm trying to figure out if the sideways thumbnails are cached somewhere or still broken.

**Comment by fisharebest** [comment]
> I can't seem to find the thumbnail cache

There is no specific cache for thumbnails - just a general-purpose filesystem cache.  `/data/cache`.

You can delete it from the control panel.

**Comment by miqrogroove** [comment]
Currently PHP Version 7.4.25 with exif, GD headers Version 2.3.0, and imagick module version 3.5.1.

**Comment by miqrogroove** [comment]
Bad rotation persists both for existing media and new uploads.

**Comment by fisharebest** [comment]
webtrees 2.0.17 uses version 2.7.0 of intervention/image (see https://github.com/fisharebest/webtrees/blob/2.0.17/composer.lock#L1437-L1438)

The file you quote is from intervention itself, and maps the version 'dev-master' onto a named development branch.
This is a common technique, to allow dependent software to automatically link to the latest version.

Although out-of-date, it won't affect us, as we don't require `dev-master`.  We require `2.5` or higher (see https://github.com/fisharebest/webtrees/blob/2.0.17/composer.json#L55)

> Currently PHP Version 7.4.25 with exif, GD headers Version 2.3.0, and imagick module version 3.5.1.

..and are you 100% certain that `imagick` is being used, and not `gd`?

You can check hereh: https://github.com/fisharebest/webtrees/blob/2.0.17/app/Factories/ImageFactory.php#L360

**Comment by miqrogroove** [comment]
Yes, it shows `imagick` as the driver in that spot.

**Comment by miqrogroove** [comment]
Trying to trace this better for you.  I have a sample image where I can get a response directly from `exif_read_data()` having Orientation = 8.  When I upload the same image to webtrees, the `$image->exif()` command always responds with Orientation = 0.

**Comment by miqrogroove** [comment]
When Intervention passes the stream to `exif_read_data()` I get Orientation = 0.  When I use `fopen()` directly and pass that stream, I get Orientation = 8.  So, yes, it's capable of reading stream data, but the stream passed by webtrees seems to be incompatible or in the wrong state.

**Comment by fisharebest** [comment]
If you upload the image to the dev server, does it work there?

**Comment by miqrogroove** [comment]
No, it is not working on the dev server.  I suggest reopening the issue or continuing in a new one.

![Screenshot_20211104-120650_Chrome](https://user-images.githubusercontent.com/1371835/140376208-ed6f30ab-f7a8-4ebd-a96a-6c9e5dbbe770.jpg)

For reference,  there are some easy to use test images at https://github.com/recurser/exif-orientation-examples

**Comment by fisharebest** [comment]
I can see that

* the code is calling `return $image->orientate();` correctly.
* the code is using the `imagick` driver.
* the image is not rotated.
* the function returns successfully.

The upstream fix was merged on 05 JUL 2021, and so should have been included in the 2.6.0 release.
webtrees is currently using 2.7.0

So, it likes like the intervention/image library isn't working as expected.

**Comment by miqrogroove** [comment]
I agree with your observations but the conclusion is wrong.  I determined the stream passed into the underlying functions does not represent the uploaded image and does not contain the exif headers.  Therefore, some or all of those functions are working as intended.  Could we drill down more and find out why the image is being modified prior to pulling the exif data?

**Comment by fisharebest** [comment]
Adding a line of debug:

```
var_dump($filesystem->fileSize($path), strlen(stream_get_contents($filesystem->readStream($path))));
$image = $this->imageManager()->make($filesystem->readStream($path));
```

gives this result

`int(352727) int(352727)`

>  I determined the stream passed into the underlying functions does not represent the uploaded image and does not contain the exif headers

They look the same to me.  What am I missing?

**Comment by miqrogroove** [comment]
Clearly the file stream is modified somewhere between the upload and the call to `orientate()`.  It's worth finding out the exact problem and whether it requires an upstream issue.  I didn't have time to debug this last night.  I can experiment with it later this week.

**Comment by elysch** [comment]
Hello.

I'm using 2.1.20 and my thumbnails sometimes get rotated still.

**Comment by fisharebest** [comment]
The fix will be availabl ein webtrees 2.2.0.  It requires an updated version of the image library, which requires an updated version of PHP.

webtrees 2.1.x needs to run on an old version of PHP (7.4), and cannot receive this fix.
